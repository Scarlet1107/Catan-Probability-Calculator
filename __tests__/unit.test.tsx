import React from "react";
import { fireEvent, render, screen, cleanup } from "@testing-library/react";
import Home from "../app/page";

// 要素が存在するかどうかなどの基本的なテスト
describe("Basic tests for the presence of elements", () => {
  beforeEach(() => {
    render(<Home />);
  });

  afterEach(() => {
    cleanup();
    localStorage.removeItem("settlements");
  });

  // 1. タイトルがある
  test("Title exists", () => {
    const titleElement = screen.getByText(/カタン確率計算アプリ/i);
    expect(titleElement).toBeInTheDocument();
  });

  // 2. ボタンが正しい数、正しい種類存在する
  test("Correct buttons exist", () => {
    const addSettlementButton = screen.getByTestId("addSettlementButton");
    const upgradeButtons = screen.getAllByTestId("upgradeButton");
    const deleteButtons = screen.getAllByTestId("deleteButton");
    expect(addSettlementButton).toBeInTheDocument();
    expect(upgradeButtons).toHaveLength(2);
    expect(deleteButtons).toHaveLength(2);
  });

  // 3. インプットタグが正しい数あり、値が正しい
  test("Input exists", () => {
    const inputs = screen.getAllByTestId("settlementNameInput");
    expect(inputs).toHaveLength(2);
    expect(inputs[0]).toHaveValue("開拓地1");
    expect(inputs[1]).toHaveValue("開拓地2");
  });

  // 4. セレクトタグが正しい数ある
  test("Select exists", () => {
    const selects = screen.getAllByRole("combobox");
    const resourceSelects = screen.getAllByTestId("resourceSelect");
    const numberSelects = screen.getAllByTestId("numberSelect");
    expect(selects).toHaveLength(12);
    expect(resourceSelects).toHaveLength(6);
    expect(numberSelects).toHaveLength(6);
  });

  // 5. オプションが正しい
  test("Options are correct", () => {
    const resourceSelects = screen.getAllByTestId("resourceSelect");
    const numberSelects = screen.getAllByTestId("numberSelect");

    const resourceOptions = ["木材", "レンガ", "小麦", "鉄", "羊毛"];
    const numberOptions = ["2", "3", "4", "5", "6", "8", "9", "10", "11", "12"];

    resourceSelects.forEach((select) => {
      const options = Array.from(select.querySelectorAll("option")).map(
        (option) => option.textContent
      );
      const isResourceOption = resourceOptions.every((option) =>
        options.includes(option)
      );
      expect(isResourceOption).toBeTruthy();
    });

    numberSelects.forEach((select) => {
      const options = Array.from(select.querySelectorAll("option")).map(
        (option) => option.textContent
      );
      const isNumberOption = numberOptions.every((option) =>
        options.includes(option)
      );
      expect(isNumberOption).toBeTruthy();
    });
  });

  // 6. 棒グラフがある
  test("Chart exists", () => {
    const chart = screen.getByRole("img");
    expect(chart).toBeInTheDocument();
  });

  // 7. 期待値が正しく初期化されている
  test("Expected value is initialized correctly", () => {
    expect(screen.getByTestId("expectedValue")).toHaveTextContent("0");
  });

  // 8. 資源取得確率が正しく初期化されている
  test("Resource probability is initialized correctly", () => {
    expect(screen.getByTestId("probability")).toHaveTextContent("0");
  });

  // 9. おすすめの数字が正しく初期化されている
  test("Recommended number is initialized correctly", () => {
    expect(screen.getByTestId("recommendedNumbers")).toHaveTextContent(
      "6, 8, 5"
    );
  });

  // 10. 都市化におすすめの開拓地が正しく初期化されている
  test("Recommended settlement is initialized correctly", () => {
    expect(screen.getByTestId("recommendedSettlementName")).toHaveTextContent(
      ""
    );
  });

  // 11. 資源のみを選択したとき、計算が実行されない
  test("Calculation does not execute when only resources are selected", () => {
    const resourceSelects = screen.getAllByTestId("resourceSelect");
    fireEvent.change(resourceSelects[0], { target: { value: "wood" } });
    fireEvent.change(resourceSelects[1], { target: { value: "bricks" } });
    fireEvent.change(resourceSelects[2], { target: { value: "wheat" } });

    expect(resourceSelects[0]).toHaveValue("wood");
    expect(resourceSelects[1]).toHaveValue("bricks");
    expect(resourceSelects[2]).toHaveValue("wheat");
    expect(screen.getByTestId("expectedValue")).toHaveTextContent("0");
    expect(screen.getByTestId("probability")).toHaveTextContent("0");
  });

  // 12. 数字のみを選択したとき、計算が実行されない
  test("Calculation does not execute when only numbers are selected", () => {
    const numberSelects = screen.getAllByTestId("numberSelect");
    fireEvent.change(numberSelects[0], { target: { value: "2" } });
    fireEvent.change(numberSelects[1], { target: { value: "6" } });
    fireEvent.change(numberSelects[2], { target: { value: "8" } });

    expect(numberSelects[0]).toHaveValue("2");
    expect(numberSelects[1]).toHaveValue("6");
    expect(numberSelects[2]).toHaveValue("8");
    expect(screen.getByTestId("expectedValue")).toHaveTextContent("0");
    expect(screen.getByTestId("probability")).toHaveTextContent("0");
  });

  // 13. スナップショットテスト
  test("renders correctly and matches snapshot", () => {
    const { asFragment } = render(<Home />);
    expect(asFragment()).toMatchSnapshot();
  });
});
