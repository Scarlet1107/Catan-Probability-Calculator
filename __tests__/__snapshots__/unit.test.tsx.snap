// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic tests for the presence of elements renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="bg-custom-bg bg-cover bg-center"
  >
    <div>
      <header
        class="bg-blue-200 p-4 flex justify-around"
      >
        <div
          class="text-3xl font-medium"
        >
          カタン確率計算アプリ
        </div>
        <div
          class="text-xl"
        >
          <button
            data-testid="changeDisplayModeToNumeric"
          >
            ランク表示に変更
          </button>
        </div>
      </header>
    </div>
    <div
      class="flex h-screen justify-center bg-white"
    >
      <div
        class="w-1/2 pl-8 pt-4 mr-4"
      >
        <div>
          開拓地の情報から、資源取得確率や期待値を計算できます。
        </div>
        <div
          class="grid grid-cols-5 gap-4 place-items-center mt-6 "
        >
          <button
            class="button bg-blue-500 hover:bg-blue-600 place-self-start w-4/5 h-5/6"
            data-testid="addSettlementButton"
            tabindex="1"
          >
            開拓地を追加
          </button>
          <div
            class="text-xl font-medium"
          >
            資源１
          </div>
          <div
            class="text-xl font-medium"
          >
            資源２
          </div>
          <div
            class="text-xl font-medium"
          >
            資源３
          </div>
          <button
            class="button mb-4 bg-gray-500 hover:bg-red-700 place-self-end w-4/5 mr-2"
            data-testid="resetButton"
            tabindex="1"
          >
            初期化
          </button>
        </div>
        <div>
          <div
            class="grid grid-cols-5 gap-4 mb-4 place-items-end"
          >
            <input
              class="border border-gray-300 rounded px-4 py-2 mb-2 h-1/2 w-4/5 place-self-start self-center"
              data-testid="settlementNameInput"
              placeholder="開拓地"
              tabindex="2"
              type="text"
              value="開拓地1"
            />
            <div
              class=""
            >
              <select
                class="justify-self-center w-4/5 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="resourceSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  value="wood"
                >
                  木材
                </option>
                <option
                  value="bricks"
                >
                  レンガ
                </option>
                <option
                  value="wheat"
                >
                  小麦
                </option>
                <option
                  value="iron"
                >
                  鉄
                </option>
                <option
                  value="wool"
                >
                  羊毛
                </option>
              </select>
              <select
                class="w-4/5 mt-2 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="numberSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  class=""
                  value="2"
                >
                  2
                </option>
                <option
                  class=""
                  value="3"
                >
                  3
                </option>
                <option
                  class=""
                  value="4"
                >
                  4
                </option>
                <option
                  class=""
                  value="5"
                >
                  5
                </option>
                <option
                  class="text-red-500"
                  value="6"
                >
                  6
                </option>
                <option
                  class="text-red-500"
                  value="8"
                >
                  8
                </option>
                <option
                  class=""
                  value="9"
                >
                  9
                </option>
                <option
                  class=""
                  value="10"
                >
                  10
                </option>
                <option
                  class=""
                  value="11"
                >
                  11
                </option>
                <option
                  class=""
                  value="12"
                >
                  12
                </option>
              </select>
            </div>
            <div
              class=""
            >
              <select
                class="justify-self-center w-4/5 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="resourceSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  value="wood"
                >
                  木材
                </option>
                <option
                  value="bricks"
                >
                  レンガ
                </option>
                <option
                  value="wheat"
                >
                  小麦
                </option>
                <option
                  value="iron"
                >
                  鉄
                </option>
                <option
                  value="wool"
                >
                  羊毛
                </option>
              </select>
              <select
                class="w-4/5 mt-2 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="numberSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  class=""
                  value="2"
                >
                  2
                </option>
                <option
                  class=""
                  value="3"
                >
                  3
                </option>
                <option
                  class=""
                  value="4"
                >
                  4
                </option>
                <option
                  class=""
                  value="5"
                >
                  5
                </option>
                <option
                  class="text-red-500"
                  value="6"
                >
                  6
                </option>
                <option
                  class="text-red-500"
                  value="8"
                >
                  8
                </option>
                <option
                  class=""
                  value="9"
                >
                  9
                </option>
                <option
                  class=""
                  value="10"
                >
                  10
                </option>
                <option
                  class=""
                  value="11"
                >
                  11
                </option>
                <option
                  class=""
                  value="12"
                >
                  12
                </option>
              </select>
            </div>
            <div
              class=""
            >
              <select
                class="justify-self-center w-4/5 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="resourceSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  value="wood"
                >
                  木材
                </option>
                <option
                  value="bricks"
                >
                  レンガ
                </option>
                <option
                  value="wheat"
                >
                  小麦
                </option>
                <option
                  value="iron"
                >
                  鉄
                </option>
                <option
                  value="wool"
                >
                  羊毛
                </option>
              </select>
              <select
                class="w-4/5 mt-2 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="numberSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  class=""
                  value="2"
                >
                  2
                </option>
                <option
                  class=""
                  value="3"
                >
                  3
                </option>
                <option
                  class=""
                  value="4"
                >
                  4
                </option>
                <option
                  class=""
                  value="5"
                >
                  5
                </option>
                <option
                  class="text-red-500"
                  value="6"
                >
                  6
                </option>
                <option
                  class="text-red-500"
                  value="8"
                >
                  8
                </option>
                <option
                  class=""
                  value="9"
                >
                  9
                </option>
                <option
                  class=""
                  value="10"
                >
                  10
                </option>
                <option
                  class=""
                  value="11"
                >
                  11
                </option>
                <option
                  class=""
                  value="12"
                >
                  12
                </option>
              </select>
            </div>
            <div
              class="flex flex-col w-full mr-2"
            >
              <button
                class="button bg-blue-500 hover:bg-blue-600"
                data-testid="upgradeButton"
                tabindex="2"
              >
                都市化
              </button>
              <button
                class="button bg-red-500 hover:bg-red-600 mt-2"
                data-testid="deleteButton"
                tabindex="2"
              >
                消
              </button>
            </div>
          </div>
          <div
            class="grid grid-cols-5 gap-4 mb-4 place-items-end"
          >
            <input
              class="border border-gray-300 rounded px-4 py-2 mb-2 h-1/2 w-4/5 place-self-start self-center"
              data-testid="settlementNameInput"
              placeholder="開拓地"
              tabindex="2"
              type="text"
              value="開拓地2"
            />
            <div
              class=""
            >
              <select
                class="justify-self-center w-4/5 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="resourceSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  value="wood"
                >
                  木材
                </option>
                <option
                  value="bricks"
                >
                  レンガ
                </option>
                <option
                  value="wheat"
                >
                  小麦
                </option>
                <option
                  value="iron"
                >
                  鉄
                </option>
                <option
                  value="wool"
                >
                  羊毛
                </option>
              </select>
              <select
                class="w-4/5 mt-2 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="numberSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  class=""
                  value="2"
                >
                  2
                </option>
                <option
                  class=""
                  value="3"
                >
                  3
                </option>
                <option
                  class=""
                  value="4"
                >
                  4
                </option>
                <option
                  class=""
                  value="5"
                >
                  5
                </option>
                <option
                  class="text-red-500"
                  value="6"
                >
                  6
                </option>
                <option
                  class="text-red-500"
                  value="8"
                >
                  8
                </option>
                <option
                  class=""
                  value="9"
                >
                  9
                </option>
                <option
                  class=""
                  value="10"
                >
                  10
                </option>
                <option
                  class=""
                  value="11"
                >
                  11
                </option>
                <option
                  class=""
                  value="12"
                >
                  12
                </option>
              </select>
            </div>
            <div
              class=""
            >
              <select
                class="justify-self-center w-4/5 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="resourceSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  value="wood"
                >
                  木材
                </option>
                <option
                  value="bricks"
                >
                  レンガ
                </option>
                <option
                  value="wheat"
                >
                  小麦
                </option>
                <option
                  value="iron"
                >
                  鉄
                </option>
                <option
                  value="wool"
                >
                  羊毛
                </option>
              </select>
              <select
                class="w-4/5 mt-2 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="numberSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  class=""
                  value="2"
                >
                  2
                </option>
                <option
                  class=""
                  value="3"
                >
                  3
                </option>
                <option
                  class=""
                  value="4"
                >
                  4
                </option>
                <option
                  class=""
                  value="5"
                >
                  5
                </option>
                <option
                  class="text-red-500"
                  value="6"
                >
                  6
                </option>
                <option
                  class="text-red-500"
                  value="8"
                >
                  8
                </option>
                <option
                  class=""
                  value="9"
                >
                  9
                </option>
                <option
                  class=""
                  value="10"
                >
                  10
                </option>
                <option
                  class=""
                  value="11"
                >
                  11
                </option>
                <option
                  class=""
                  value="12"
                >
                  12
                </option>
              </select>
            </div>
            <div
              class=""
            >
              <select
                class="justify-self-center w-4/5 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="resourceSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  value="wood"
                >
                  木材
                </option>
                <option
                  value="bricks"
                >
                  レンガ
                </option>
                <option
                  value="wheat"
                >
                  小麦
                </option>
                <option
                  value="iron"
                >
                  鉄
                </option>
                <option
                  value="wool"
                >
                  羊毛
                </option>
              </select>
              <select
                class="w-4/5 mt-2 border border-gray-300 rounded px-4 py-2 focus:border-blue-500 focus:ring focus:ring-blue-500 focus:ring-opacity-50"
                data-testid="numberSelect"
                tabindex="1"
              >
                <option
                  value=""
                />
                <option
                  class=""
                  value="2"
                >
                  2
                </option>
                <option
                  class=""
                  value="3"
                >
                  3
                </option>
                <option
                  class=""
                  value="4"
                >
                  4
                </option>
                <option
                  class=""
                  value="5"
                >
                  5
                </option>
                <option
                  class="text-red-500"
                  value="6"
                >
                  6
                </option>
                <option
                  class="text-red-500"
                  value="8"
                >
                  8
                </option>
                <option
                  class=""
                  value="9"
                >
                  9
                </option>
                <option
                  class=""
                  value="10"
                >
                  10
                </option>
                <option
                  class=""
                  value="11"
                >
                  11
                </option>
                <option
                  class=""
                  value="12"
                >
                  12
                </option>
              </select>
            </div>
            <div
              class="flex flex-col w-full mr-2"
            >
              <button
                class="button bg-blue-500 hover:bg-blue-600"
                data-testid="upgradeButton"
                tabindex="2"
              >
                都市化
              </button>
              <button
                class="button bg-red-500 hover:bg-red-600 mt-2"
                data-testid="deleteButton"
                tabindex="2"
              >
                消
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="w-1/2 ml-4 mr-8"
      >
        <div
          class="text-2xl m-4"
        >
          <span
            class="relative inline-block cursor-pointer"
          >
            期待値
            <span
              class="opacity-0 absolute top-full left-1/2 transform -translate-x-1/2 mt-2 w-64 bg-gray-800 text-white text-center rounded py-2 transition-opacity duration-300 pointer-events-none tooltip"
            >
              2から12の36通りの数字が出たときに、何枚の資源がもらえるかを表します
              <span
                class="tooltip-arrow"
              />
            </span>
          </span>
           : 
          <span
            data-testid="expectedValue"
          >
            0
          </span>
        </div>
        <div
          class="text-2xl m-4"
        >
          <span
            class="relative inline-block cursor-pointer"
          >
            資源取得確率
            <span
              class="opacity-0 absolute top-full left-1/2 transform -translate-x-1/2 mt-2 w-64 bg-gray-800 text-white text-center rounded py-2 transition-opacity duration-300 pointer-events-none tooltip"
            >
              ダイスを振ったときに資源が１枚以上もらえる確率を表します
              <span
                class="tooltip-arrow"
              />
            </span>
          </span>
           : 
          <span
            data-testid="probability"
          >
            0 / 36
          </span>
        </div>
        <div
          class="text-2xl m-4"
        >
          <span
            class="relative inline-block cursor-pointer"
          >
            おすすめの数字
            <span
              class="opacity-0 absolute top-full left-1/2 transform -translate-x-1/2 mt-2 w-64 bg-gray-800 text-white text-center rounded py-2 transition-opacity duration-300 pointer-events-none tooltip"
            >
              おすすめの開拓地の数字の場所を表します
              <span
                class="tooltip-arrow"
              />
            </span>
          </span>
           : 
          <span
            data-testid="recommendedNumbers"
          >
            6, 8, 5
          </span>
        </div>
        <div
          class="text-2xl m-4"
        >
          <span
            class="relative inline-block cursor-pointer"
          >
            アップグレードにおすすめの開拓地
            <span
              class="opacity-0 absolute top-full left-1/2 transform -translate-x-1/2 mt-2 w-64 bg-gray-800 text-white text-center rounded py-2 transition-opacity duration-300 pointer-events-none tooltip"
            >
              資源取得量からアップグレードにおすすめの開拓地を計算します。資源取得量のみを考慮しているのでお気をつけください
              <span
                class="tooltip-arrow"
              />
            </span>
          </span>
           : 「
          <span
            data-testid="recommendedSettlementName"
          />
          」
        </div>
        <div
          class="text-2xl flex justify-center"
        >
          <canvas
            height="0"
            role="img"
            style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
            width="0"
          />
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
